internal/controllers/machine/machine_controller.go

```diff
@@ -1025,3 +1088,193 @@ func (r *Reconciler) nodeToMachine(ctx context.Context, o client.Object) []recon
+			// Skip VolumeAttachments which are not in attached state.
+			if !volumeAttachment.Status.Attached {
+				continue
+			}
+			// Skip VolumeAttachments which are not for the given node.
+			if volumeAttachment.Spec.NodeName != nodeName {
+				continue
+			}
+			volumeAttachments = append(volumeAttachments, &volumeAttachment)
+		}
+
+		if volumeAttachmentList.GetContinue() == "" {
+			break
+		}
+	}
+
+	return volumeAttachments, nil
+}
+
+// getPersistentVolumesWaitingForDetach returns a list of all persistentVolumes which either have
+// the calculated AttachedVolume name in attachedVolumeNames or their name in attachedPVNames.
+// PersistentVolumes which refer a PersistentVolumeClaim contained in pvcsToIgnore are filtered out.
+// If there are names in attachedVolumeNames or attachedPVNames without a corresponding PV, this returns an error.
+func getPersistentVolumesWaitingForDetach(ctx context.Context, c client.Client, attachedVolumeNames, attachedPVNames, pvcsToIgnore sets.Set[string]) ([]*corev1.PersistentVolume, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	pvsWaitingForDetach := []*corev1.PersistentVolume{}
+	foundAttachedVolumeNames := sets.Set[string]{}
+	foundAttachedPVNames := sets.Set[string]{}
+
+	// List all PersistentVolumes and preserve the ones we have to wait for.
+	// Also store the found VolumeHandles and names of PersistentVolumes to check
+	// that all expected PersistentVolumes have been found.
+	// Note: pvsWaitingFor will not include PersistentVolumes we want to ignore.
+	// These are usually volumes which have a Pod which is expected to be kept.
+	persistentVolumeList := &corev1.PersistentVolumeList{}
+	for {
+		listOpts := []client.ListOption{
+			client.Continue(persistentVolumeList.GetContinue()),
+			client.Limit(100),
+		}
+		if err := c.List(ctx, persistentVolumeList, listOpts...); err != nil {
+			return nil, errors.Wrapf(err, "failed to list %T", persistentVolumeList)
+		}
+
+		for _, persistentVolume := range persistentVolumeList.Items {
+			found := false
+			// Lookup if the PersistentVolume matches an entry in attachedVolumeNames.
+			if persistentVolume.Spec.CSI != nil {
+				attachedVolumeName := fmt.Sprintf("kubernetes.io/csi/%s^%s", persistentVolume.Spec.CSI.Driver, persistentVolume.Spec.CSI.VolumeHandle)
+				if attachedVolumeNames.Has(attachedVolumeName) {
+					foundAttachedVolumeNames.Insert(attachedVolumeName)
+					found = true
+				}
+			}
+
+			// Lookup if the PersistentVolume matches an entry in attachedPVNames.
+			if attachedPVNames.Has(persistentVolume.Name) {
+				foundAttachedPVNames.Insert(persistentVolume.Name)
+				found = true
+			}
+
+			// PersistentVolume which do not match an entry in attachedVolumeNames or
+			// attachedPVNames can be ignored.
+			if !found {
+				continue
+			}
+
+			// Ignore PersistentVolumes which refer a PersistentVolumeClaim from pvcsToIgnore.
+			if persistentVolume.Spec.ClaimRef != nil && persistentVolume.Spec.ClaimRef.Kind == "PersistentVolumeClaim" {
```

Comment:
Maybe it's slightly easier to just put the PVCs into a "attachedPVCs" list and then do a attachedPVCs.Difference(pvcsToIgnore) at the end of this func